File: frontend/src/api/compare.ts
import { api } from './client'

export interface DiffItem {
  type: 'equal' | 'add' | 'del'
  text: string
}

export interface CompareResponse {
  diff: DiffItem[]
}

export interface DiffResponse {
  original: string
  modified: string
  diff: DiffItem[]
}

export interface CompareMetaFramework {
  has_diff: boolean
}

export interface CompareMetaChapterItem {
  number: number
  has_diff: boolean
  label: string
}

export interface CompareMetaResponse {
  has_any: boolean
  framework: CompareMetaFramework
  chapters: CompareMetaChapterItem[]
}

export async function postCompare(
  original: string,
  modified: string,
): Promise<CompareResponse> {
  const { data } = await api.post<CompareResponse>('/api/compare', {
    original,
    modified,
  })
  return data
}

export async function getFrameworkDiff(taskId: string): Promise<DiffResponse> {
  const { data } = await api.get<DiffResponse>(
    `/api/tasks/${taskId}/steps/framework/diff`,
  )
  return data
}

export async function getChaptersDiff(
  taskId: string,
  chapterNumber: number,
): Promise<DiffResponse> {
  const { data } = await api.get<DiffResponse>(
    `/api/tasks/${taskId}/steps/chapters/diff`,
    { params: { chapter_number: chapterNumber } },
  )
  return data
}

export async function getTaskCompareMeta(taskId: string): Promise<CompareMetaResponse> {
  const { data } = await api.get<CompareMetaResponse>(`/api/tasks/${taskId}/compare-meta`)
  return data
}

File: backend/app/schemas/compare.py
"""Pydantic schemas for compare (diff) API."""
from typing import Literal

from pydantic import BaseModel


class CompareRequest(BaseModel):
    """Request body for POST /api/compare."""

    original: str = ""
    modified: str = ""


class DiffItem(BaseModel):
    """One segment of a structured diff (equal=unchanged, del=deleted, add=added)."""

    type: Literal["equal", "add", "del"]
    text: str


class CompareResponse(BaseModel):
    """Response for POST /api/compare."""

    diff: list[DiffItem]


class DiffResponse(BaseModel):
    """Response for GET .../steps/framework/diff and .../steps/chapters/diff."""

    original: str
    modified: str
    diff: list[DiffItem]


class FrameworkCompareMeta(BaseModel):
    """Metadata for framework diff availability."""

    has_diff: bool


class ChapterCompareMetaItem(BaseModel):
    """One chapter entry in compare meta."""

    number: int
    has_diff: bool
    label: str


class CompareMetaResponse(BaseModel):
    """Task-level compare metadata: which items have before/after versions."""

    has_any: bool
    framework: FrameworkCompareMeta
    chapters: list[ChapterCompareMetaItem]
